### DISCUSSION TITLE
Server certificate expires - how to renew using pivpn?

### DISCUSSION THREAD

**Discussion by heiko-wolf** [pivpn/pivpn#1560] [Support]
Hi there,
just seen that a server certificate created while installing pivpn is expiring the next days.
How to renew this on a pivpn setup without the need of updating all client configs?
Thanks for any hint on this !!
best regards,
Heiko

**Comment by orazioedoardo** [comment]
~~Just revoke the client and create a new one with the same name (you will need to copy the .ovpn file again).~~

**Reply by heiko-wolf** [reply]
thanks for your hint .. but in this case it is not a client certificate which has expired but the server certificate itself.
That is what was created while installing pivpn a year before:

: NOTE : The first entry should always be your valid server!

::: Certificate Status List :::
 ::  Status  ||   Name   :: 
     Valid   ::   server_kYtAVzcmkMC9efYZ

But now this server cert has timed out and I cant find a way to just revoke it and create a new one using the pivpn command.
Will try to do this using easy_rsa but a better way was using pivpn for this.

**Reply by orazioedoardo** [reply]
Sorry, was in hurry and just assumed it was an expired client.

If the CA (certificate authority) is not about to expire too, you can generate a new server certificate signed by the CA, check the `Validity`: `Not After` field in the output of the following command (but don't post the output) `openssl x509 -text < /etc/openvpn/easy-rsa/pki/ca.crt`.

To generate a new certificate, get root, enter `/etc/openvpn/easy-rsa` and run `EASYRSA_CERT_EXPIRE=3650 ./easyrsa build-server-full "$(hostname -s)_$(</proc/sys/kernel/random/uuid)" nopass`. This will generate a new server certificate, valid for 10 years (3650 days) named with the format `hostname_randomuuid`.

Next, open `/etc/openvpn/server.conf` and change `server_kYtAVzcmkMC9efYZ` in the lines with `cert` and `key` to the name of the new server certificate. Restart the server with `systemctl restart openvpn` and try to reconnect.

**Reply by heiko-wolf** [reply]
Hey, thanks for your fast response to this one :)
Have tried this already but with no luck in the first round ... now I found out that the configured clients have the "verify-x509-name" option enabled in their ovpn files which results in the need to update them all before they are able to use the new server cert.

To prevent this I have to revoke and delete the expired server cert as I understand and create a new one with the same name.

Is this possible or do I have to update alle client configs too?

I am wondering why the default setup of pivpn creates a CA which lasts for 10 years and the server cert is just valid for one year.

**Reply by orazioedoardo** [reply]
>  I found out that the configured clients have the "verify-x509-name"

True! Well, as you said you can revoke - delete - generate the new server certificate. Backup the `/etc/openvpn/easy-rsa` folder first.

To revoke, simply run `./easyrsa revoke server_kYtAVzcmkMC9efYZ`. Then delete the `.req`, `.key` and `.crt` files named after the server in the `pki/reqs`, `pki/private` and `pki/isssued` subfolders. Next, update the CRL (certificate revocation list) with `./easyrsa gen-crl` and copy that to a location where openvpn will find it ` cp /etc/openvpn/easy-rsa/pki/crl.pem /etc/openvpn/crl.pem`.

You can now create a new certificate with the same name using `EASYRSA_CERT_EXPIRE=3650 ./easyrsa build-server-full server_kYtAVzcmkMC9efYZ nopass`. You may also want to go through the same steps to revoke and delete the previous `hostname_randomuuid` server certificate.

> I am wondering why the default setup of pivpn creates a CA which lasts for 10 years and the server cert is just valid for one year.

At least 2 years ago, PiVPN has been updated to create server certificates valid for 10 years.

**Reply by heiko-wolf** [reply]
worked like a charm ... many thanks for this fast and really perfect support :) 

Just a last question .. I learned now that I use this pivpn instance since nearly three years. Working perfectly all the time.
As I am keeping the OS updated I now wonder how to update the pivpn stuff itself?!? .. do I have to reinstall and import the CA or is there a way to update using "pivpn --check-update" or something else ?

**Reply by orazioedoardo** [reply]
PiVPN did have an update feature that downloaded the most recent scripts but it has been disabled because newer scripts also expect different configurations and directory structures. Since PiVPN installs on openvpn and wireguard, updating the system does most of the work so basically you don't update PiVPN. There is also a wiki page from one of the original mantainers: https://github.com/pivpn/pivpn/wiki/Upgrading#how-do-we-upgrade-pivpn.

**Reply by coolapso** [reply]
@orazioedoardo should we add/create a feature request issue for a "Renew server certificate" option?

**Comment by TinCanTech** [comment]
I would **not** recommend using EasyRSA `renew` until version `3.1`

Simply `revoke foo` the old cert with EasyRSA and build a new one from scratch `build-server-full foo`.

**Reply by coolapso** [reply]
hence the renew between "" on my suggestion. when I say providing it as an option through pivpn scripts its as an abstraction method for the whole procedure explained above by @orazioedoardo

**Reply by TinCanTech** [reply]
The problem with the explanation given is that the server X509 name will change and that means all the clients have to change the line `verify-x509-name old-server-name` to `new-server-name`.  EasyRSA `renew` does not cause this knock on effect, however, versions of EasyRSA prior to `3.1` all leave the old server certificate unrevoked.

**Reply by orazioedoardo** [reply]
> The problem with the explanation given is that the server X509 name will change and that means all the clients have to change the line

Wouldn't be enough to recreate the certificate with the same name? Instructions in [this comment](https://github.com/pivpn/pivpn/discussions/1560#discussioncomment-3077140) worked according to @heiko-wolf

**Reply by TinCanTech** [reply]
> > The problem with the explanation given is that the server X509 name will change and that means all the clients have to change the line
> 
> Wouldn't be enough to recreate the certificate with the same name? Instructions in [this comment](https://github.com/pivpn/pivpn/discussions/1560#discussioncomment-3077140) worked according to @heiko-wolf

Yes, that would work. Except that, when you use `easyrsa revoke foo`, all the files which you mention to delete are already moved to the `revoked` folder, so they do not need to be deleted at all.
